### PR TITLE
Add sidebar enabled experiment metrics

### PR DESCRIPTION
**Pull request by DJAndries**
Resolves https://github.com/brave/brave-browser/issues/36815

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

For "A" metrics, test using `--enable-features="SidebarShowAlwaysOnStable:open_one_shot_leo_panel/false"` switch. For "B" metrics, test using `--enable-features="SidebarShowAlwaysOnStable:open_one_shot_leo_panel/true"` switch.

For `Brave.Core.SidebarSettings.SidebarEnabledA/B`:

0. Start with fresh profile, ensure one metric only exists if using one of the above switches, with the value 0.
1. Change the sidebar setting, ensure metric updates accordingly.

For `Brave.AIChat.UsageWeekly.SidebarEnabledA/B`, `Brave.AIChat.UsageDaily.SidebarEnabledA/B` and `Brave.AIChat.Enabled.SidebarEnabledA/B`:

0. Start with fresh profile with no switches, ensure no metric similar to `Brave.AIChat.UsageWeekly` or `Brave.AIChat.UsageDaily.2` or `Brave.AIChat.Enabled.2` exists.
1. Use Leo, ensure `Brave.AIChat.UsageWeekly`, `Brave.AIChat.UsageDaily.2` and `Brave.AIChat.Enabled.2` exists with the correct value.
2. Start with fresh profile using one of the switches, ensure no relevant metrics exist.
3. Use Leo, ensure  `Brave.AIChat.UsageWeekly.SidebarEnabledA/B`, `Brave.AIChat.UsageDaily.SidebarEnabledA/B`, and `Brave.AIChat.Enabled.SidebarEnabledA/B`  exists with the correct value.

### CODE REVIEW

**Comment by iefremov** [comment]
why do you need the browser/common split? This is commonly used when there is a separation between browser/renderer code, in this case there should be browser, renderer and common subdirs. If there is no renderer, then this extra layer of folders is redundant I'd say

File: components/ai_chat/core/browser/ai_chat_metrics.cc [hunk 1]
```diff
@@ -29,6 +34,65 @@ const int kOmniboxOpenBuckets[] = {-1, 0, 3, 5, 10, 25};
...
+
+    // If the mode applies for a given histogram name, report it as usual.
+    // If not, do not report & suspend metric, so we don't double count
+    // reporting two or more metrics.
+    int report_value = current_mode == i_mode ? value : INT_MAX - 1;
+    base::UmaHistogramExactLinear(histogram_name, report_value, 3);
+  }
+}
+
+const char* GetEnabledHistogramName(SidebarDefaultMode mode) {
```
Comment (iefremov) [inline components/ai_chat/core/browser/ai_chat_metrics.cc:54]: i would use constexpr flatmaps instead

**Comment by DJAndries** [comment]
> why do you need the browser/common split? This is commonly used when there is a separation between browser/renderer code, in this case there should be browser, renderer and common subdirs. If there is no renderer, then this extra layer of folders is redundant I'd say

`ai_chat_metrics.cc` in `//components/ai_chat/core/browser` must access the features that were previously in `//components/sidebar/browser`. Since `//components/sidebar/browser` depends on `//components/ai_chat/core/browser`, a new module is required to avoid a circular dependency.

**Review by supermassive — APPROVED** [review]
wallet core lgtm

**Review by iefremov — APPROVED** [review]

File: chromium_src/chrome/browser/sources.gni [hunk 2]
```diff
@@ -23,5 +23,6 @@ if (enable_brave_vpn) {
 }
 
 if (toolkit_views) {
-  brave_chromium_src_chrome_browser_deps += [ "//brave/components/sidebar" ]
+  brave_chromium_src_chrome_browser_deps +=
```
Comment (bridiver) [inline chromium_src/chrome/browser/sources.gni:26]: why are these here instead of `//brave/browser/sources.gni`? I know this file wasn't created in this PR (it was a long time ago), but what is the criteria for inclusion here?

**Comment by GeetaSarvadnya** [comment]
Verification INPROGRESS on

```
Brave | 1.67.1 Chromium: 124.0.6367.29 (Official Build) nightly (64-bit)
-- | --
Revision | f4edcfc3cafa0d8609db03da09eb8582d83d1a27
OS | Windows 10 Version 22H2 (Build 19045.4170)
```

**Comment by GeetaSarvadnya** [comment]
@DJAndries I don't see the metrics `Brave.Core.SidebarSettings.SidebarEnabledA` on the local state file when I enabled the sidebar settings/change the sidebar settings to "On mouse over" or "Never". But I do see the metrics `Brave.Sidebar.SettingChange.SidebarEnabledA` in the local state file and it this metrics gives the correct response when I change the sidebar settings to "On mouse over" or "Never". Looks like the metrics name chnage in the issues description and test plan, could you look into this and let me know which one ic correct? (`Brave.Core.SidebarSettings.SidebarEnabledA` OR `Brave.Sidebar.SettingChange.SidebarEnabledA`?)

**Comment by DJAndries** [comment]
@GeetaSarvadnya sorry for the confusion, `Brave.Sidebar.SettingChange.SidebarEnabledA/B` is correct
